### DISCUSSION TITLE
How to call main.js functions in modern theme?

### DISCUSSION THREAD

**Discussion by glopesdev** [dotnet/docfx#9414] [Q&A] [answered]
I am using the new modern theme on DocFX 2.73.1. For some reason I cannot get the `main.js` functions in my template to get called. I am porting this from a site on the default template but it is not an issue with the template folder structure itself, I have checked and it's placed on `public/main.js`.

Copying the default template example works and the icon link attributes do get assigned, but how exactly should we call javascript functions on document load? For example, if you wanted to use jQuery to insert some DOM elements how would it be done now?

I found it very odd that the link attributes get assigned in the example but inspecting the resulting HTML I can't seem to find a reference to `main.js` at all? Does it get compressed and "compiled" away somewhere?

**Comment by glopesdev** [comment]
Ok, it's been a while since I came back to this level of DocFX customization, but finally figured it out. To understand I started hunting down commits referencing `main.js` and this one pointed me in the right direction: https://github.com/dotnet/docfx/commit/34c41e11ea49d88f8d8a88a75e35d7114c242e30

So if I get this straight, `main.js` is now not simply executed anymore but is somehow expected to return a module with options, and some functions in those modules are called asynchronously at appropriate times when the page is rendered.

Which functions exactly are called, their names and signatures seem to be changing a lot. Is there a plan to stabilize on this API so we can know where exactly we should hook into? I understand the idea of having specific callbacks to configure specific modules, but would be nice to have something more general like a `start()` or some kind of main function that we could define in options and have predictable expectations of when it will be called.

For now I am hooking into `configureHljs` but it feels like a hack to me, since what I want to do has nothing to do with highlight.js.

**Reply by glopesdev** [reply]
A follow-up question that I am still scratching my head around is how exactly do we insert other JS scripts using partials now? The architecture seems to have changed so now everything is built dynamically but it's very unclear where the customization entry points are.

**Reply by glopesdev** [reply, marked as answer]
Ok, finally understand that `main.js` simply runs already in the right place to manipulate DOM elements, so if you have a function you need to call, the following seems to work well:

```js
async function my_init() {
}

export default { }
my_init();
```

**Reply by yufeih** [reply]
Seems like a `start()` hook would make it more obvious
